cyw43/src/control.rs

```diff
@@ -316,6 +322,50 @@ impl<'a> Control<'a> {
         self.set_iovar_u32x2("bss", 0, 1).await; // bss = BSS_UP
     }
 
+    /// Add specified address to the list of hardware addresses the device
+    /// listens on. The address must be a Group address (I/G bit set). Up
+    /// to 10 addresses are supported by the firmware. Returns the number of
+    /// address slots filled after adding, or an error.
+    pub async fn add_multicast_address(&mut self, address: [u8; 6]) -> Result<usize, AddMulticastAddressError> {
+        // The firmware seems to ignore non-multicast addresses, so let's
+        // prevent the user from adding them and wasting space.
+        if address[0] & 0x01 != 1 {
+            return Err(AddMulticastAddressError::NotMulticast);
+        }
+
+        let mut buf = [0; 64];
+        self.get_iovar("mcast_list", &mut buf).await;
+
+        let n = u32::from_le_bytes(buf[..4].try_into().unwrap()) as usize;
+        let (slots, _) = buf[4..].as_chunks_mut();
+        let (used, free) = slots.split_at_mut(n);
+
+        if used.contains(&address) {
+            return Ok(n);
+        }
+
+        if free.is_empty() {
+            return Err(AddMulticastAddressError::NoFreeSlots);
+        }
+
+        free[0] = address;
+        let n = n + 1;
+        buf[..4].copy_from_slice(&(n as u32).to_le_bytes());
+
+        self.set_iovar_v::<80>("mcast_list", &buf).await;
+        Ok(n)
+    }
+
+    /// Retrieve the list of configured multicast hardware addresses.
+    pub async fn list_mulistcast_addresses<'b>(&mut self, buf: &'b mut [u8; 64]) -> &'b [[u8; 6]] {
```

Comment:
I'm not sure about this pattern of using a caller-provided buffer and returning a typed slice inside it. However, without `heapless::Vec` or something equivalent, I don't see a better solution.

Reply:
it's a bit strange yes, especially changing the type. in these cases it's more common to take a buffer and return how many slots have been filled:

`pub async fn list_mulistcast_addresses(&mut self, addresses: &mut [[u8; 6]; 10]) -> usize`

Reply:
Thing is I need spaces in front of the addresses for the 4 bytes length. Maybe we can just pull in `heapless`, as it's a dependency of `embassy-net` anyway? I think that would offer the nicest API.

Reply:
hmm perhaps read to a stack buffer and copy the interesting part to the user buffer then? it uses 2x the memory, but you need that anyway if you return a heapless::Vec instead. Also, the compiler tends to be dumb about returning big things by value, also causing multiple copies of it on the stack.